util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,461 @@
+		normalizedPlanMap: tsr.normalizedPlanMap.Load().(*sync.Map),
+	}
+
+	// Reset data for next report.
+	tsr.topSQLMap = make(map[string]*topSQLDataPoints)
+	tsr.normalizedSQLMap.Store(&sync.Map{})
+	tsr.normalizedPlanMap.Store(&sync.Map{})
+
+	// Send to report channel
+	select {
+	case tsr.reportDataChan <- data:
+	default:
+	}
+}
+
+type reportData struct {
+	topSQL            map[string]*topSQLDataPoints
+	normalizedSQLMap  *sync.Map
+	normalizedPlanMap *sync.Map
+}
+
+// RegisterSQL registers a normalized SQL string to a SQL digest.
+//
+// Note that the normalized SQL string can be of >1M long.
+// This function should be thread-safe, which means parallelly calling it in several goroutines should be fine.
+// It should also return immediately, and do any CPU-intensive job asynchronously.
+// TODO: benchmark test concurrent performance
+func (tsr *RemoteTopSQLReporter) RegisterSQL(sqlDigest []byte, normalizedSQL string) {
+	m := tsr.normalizedSQLMap.Load().(*sync.Map)
+	key := string(sqlDigest)
+	_, ok := m.Load(key)
+	if ok {
+		return
+	}
+	m.Store(key, normalizedSQL)
+}
+
+// RegisterPlan is like RegisterSQL, but for normalized plan strings.
+func (tsr *RemoteTopSQLReporter) RegisterPlan(planDigest []byte, normalizedPlan string) {
+	m := tsr.normalizedPlanMap.Load().(*sync.Map)
+	key := string(planDigest)
+	_, ok := m.Load(key)
+	if ok {
+		return
+	}
+	m.Store(key, normalizedPlan)
+}
+
+// prepareReportData will collect the current snapshot of data for transmission, and clear the records map.
+// This could run in parallel with `Collect()`, so we should guard it by a mutex.
+//
+// NOTE: we could optimize this using a mutex protected pointer-swapping operation,
+// which means we maintain 2 *struct which contains the maps, and snapshot() atomically
+// swaps the pointer. After this, the writing is shifted to the new struct, and
+// we can do the snapshot in the background.
+func (tsr *RemoteTopSQLReporter) prepareReportData(data reportData) []*tipb.CollectCPUTimeRequest {
+	// wait latest register finish, use sleep instead of other method to avoid performance issue in hot code path.
+	time.Sleep(time.Millisecond * 100)
+	data.normalizedPlanMap.Range(func(key, value interface{}) bool {
+		planDecoded, err := tsr.planBinaryDecoder(value.(string))
+		if err != nil {
+			logutil.BgLogger().Warn("[top-sql] decode plan failed", zap.Error(err))
+			return true
+		}
+		data.normalizedPlanMap.Store(key, planDecoded)
+		return true
+	})
+
+	i := 0
+	batch := make([]*tipb.CollectCPUTimeRequest, len(data.topSQL))
+	for _, value := range data.topSQL {
+		normalizedSQL, sqlOk := data.normalizedSQLMap.Load(string(value.SQLDigest))
+		normalizedPlan, planOk := data.normalizedPlanMap.Load(string(value.PlanDigest))
+		req := &tipb.CollectCPUTimeRequest{
+			TimestampList: value.TimestampList,
+			CpuTimeMsList: value.CPUTimeMsList,
+			SqlDigest:     value.SQLDigest,
+			PlanDigest:    value.PlanDigest,
+		}
+		if sqlOk {
```

Comment:
Please use our latest protobuf, which does not need sql text to present when uploading the time series data.